User-assigned identity flow (#44718)
dotnet/docs: docs/azure/sdk/authentication/user-assigned-managed-identity.md

---
title: Authenticate Azure-hosted .NET apps to Azure resources using a user-assigned managed identity

> description: Learn how to authenticate Azure-hosted .NET apps to other Azure services using a user-assigned identity

Reviewer: ```suggestion
description: Learn how to authenticate Azure-hosted .NET apps to other Azure services using a user-assigned managed identity.
```

ms.topic: how-to
ms.custom: devx-track-dotnet, engagement-fy23, devx-track-azurecli
ms.date: 02/06/2025
---

# Authenticate Azure-hosted .NET apps to Azure resources using a user-assigned managed identity

The recommended approach to authenticate an Azure-hosted app to other Azure resources is to use a [managed identity](/entra/identity/managed-identities-azure-resources/overview). This approach is [supported for most Azure services](/entra/identity/managed-identities-azure-resources/managed-identities-status), including apps hosted on Azure App Service, Azure Container Apps, and Azure Virtual Machines. Discover more about different authentication techniques and approaches on the [authentication overview](/dotnet/azure/sdk/authentication) page. In the sections ahead, you'll learn:

- Essential managed identity concepts
- How to create a user-assigned managed identity for your app
- How to assign roles to the user-assigned managed identity
- How to authenticate using the user-assigned managed identity from your app code

[!INCLUDE [managed-identity-concepts](../includes/managed-identity-concepts.md)]

## Create a user-assigned managed identity

> User-assigned identities are created as standalone resources in your Azure subscription using the Azure portal or the Azure CLI. Azure CLI commands can be run in the [Azure Cloud Shell](https://shell.azure.com) or on a workstation with the [Azure CLI installed](/cli/azure/install-azure-cli).

Reviewer: ```suggestion
User-assigned managed identities are created as standalone resources in your Azure subscription using the Azure portal or the Azure CLI. Azure CLI commands can be run in the [Azure Cloud Shell](https://shell.azure.com) or on a workstation with the [Azure CLI installed](/cli/azure/install-azure-cli).
```

### [Azure portal](#tab/azure-portal)

1. In the Azure portal, enter *Managed identities* in the main search bar and select the matching result under the **Services** section.
1. On the **Managed Identities** page, select **+ Create**.

    :::image type="content" source="../media/user-assigned-identity-create.png" alt-text="A screenshot showing the page to manage user-assigned identities.":::

> 1. On the **Create User Assigned Managed Identity** page, select a subscription, resource group, and region for the user-assigned identity, and then provide a name.

Reviewer: ```suggestion
1. On the **Create User Assigned Managed Identity** page, select a subscription, resource group, and region for the user-assigned managed identity, and then provide a name.
```

1. Select **Review + create** to review and validate your inputs.

>     :::image type="content" source="../media/user-assigned-identity-form.png" alt-text="A screenshot showing the form to create a user-assigned identity.":::

Reviewer: ```suggestion
    :::image type="content" source="../media/user-assigned-identity-form.png" alt-text="A screenshot showing the form to create a user-assigned managed identity.":::
```

> 1. Select **Create** to create the user-assigned identity.

Reviewer: ```suggestion
1. Select **Create** to create the user-assigned managed identity.
```

1. After the identity is created, select **Go to resource**.
1. On the new identity's **Overview** page, copy the `Client ID` value to use for later when you configure the application code.

### [Azure CLI](#tab/azure-cli)

Use the Azure CLI command [`az identity create`](/cli/azure/identity?view=azure-cli-latest#az-identity-create) to create a managed identity:

```azurecli

> az identity create --resource-group <resource-group-name> --name <identity-name>

Reviewer: ```suggestion
az identity create \
    --resource-group <resource-group-name> \
    --name <identity-name>
```

    :::image type="content" source="../media/user-assigned-identity-create.png" alt-text="A screenshot showing the page to manage user-assigned identities.":::

1. On the **Create User Assigned Managed Identity** page, select a subscription, resource group, and region for the user-assigned managed identity, and then provide a name.
1. Select **Review + create** to review and validate your inputs.

    :::image type="content" source="../media/user-assigned-identity-form.png" alt-text="A screenshot showing the form to create a user-assigned managed identity.":::

1. Select **Create** to create the user-assigned managed identity.
1. After the identity is created, select **Go to resource**.
1. On the new identity's **Overview** page, copy the **Client ID** value to use for later when you configure the application code.

### [Azure CLI](#tab/azure-cli)

Use the Azure CLI command [`az identity create`](/cli/azure/identity?view=azure-cli-latest#az-identity-create) to create a managed identity:

```azurecli
az identity create \
    --resource-group <resource-group-name> \
    --name <identity-name> \
    --query 'clientId' \
    --output json
```

> The command output prints the **ClientID** value of the created user-assigned managed identity. The **ClientID** is used to configure application code that relies on the identity.

Reviewer: ```suggestion
The command output prints the client ID of the created user-assigned managed identity. The client ID is used to configure application code that relies on the identity.
```

### [Azure portal](#tab/azure-portal)

1. In the Azure portal, navigate to the resource that hosts your app code, such as an Azure App Service or Azure Container App instance.
1. From the resource's **Overview** page, expand **Settings** and select **Identity** from the navigation.
1. On the **Identity** page, switch to the **User assigned** tab.
1. Select **+ Add** to open the **Add user assigned managed identity** panel.
1. On the **Add user assigned managed identity** panel, use the **Subscription** dropdown to filter the search results for your identities. Use the **User assigned managed identities** search box to locate the user-assigned managed identity you enabled for the Azure resource hosting your app.
1. Select the identity and choose **Add** at the bottom of the panel to continue.

    :::image type="content" source="../media/add-user-assigned-identity-to-app.png" alt-text="A screenshot showing how to associate a user-assigned identity with an app.":::

### [Azure CLI](#tab/azure-cli)

> The Azure CLI provides different commands to assign a user-assigned identity to different types of hosting services.

Reviewer: ```suggestion
The Azure CLI provides different commands to assign a user-assigned managed identity to different types of hosting services.
```

> To assign a user-assigned identity to a resource such as an Azure App Service web app using the Azure CLI, you'll need the resource ID of the identity. Use the [`az identity show`](/cli/azure/identity?view=azure-cli-latest#az-identity-show) command to retrieve the resource ID:

Reviewer: ```suggestion
To assign a user-assigned managed identity to a resource such as an Azure App Service web app using the Azure CLI, you'll need the resource ID of the identity. Use the [`az identity show`](/cli/azure/identity?view=azure-cli-latest#az-identity-show) command to retrieve the resource ID:
```

```azurecli
az identity show \
    --resource-group <your-resource-group> \
    --name <your-managed-identity-name> \
    --output json \
    --query id
```

Once you have the resource ID, use the Azure CLI command `az <resourceType> identity assign` command to associate the user-assigned identity with different resources, such as the following:

For Azure App Service, use the Azure CLI command [`az webapp identity assign`](/cli/azure/webapp/identity?view=azure-cli-latest#az-webapp-identity-assign):

```azurecli
az webapp identity assign \
    --resource-group <resource-group-name> \
    --name <webapp-name> \
    --identities <user-assigned-identity-resource-id>
```

For Azure Container Apps, use the Azure CLI command [`az containerapp identity assign`](/cli/azure/containerapp/identity?view=azure-cli-latest#az-containerapp-identity-assign):

```azurecli

> az webapp identity assign \

Reviewer: This command needs to be updated for ACA

1. Select the role from the results and then choose **Next** to move to the **Members** tab.
1. For the **Assign access to** option, select **Managed identity**.
1. For the **Members** option, choose **+ Select members** to open the **Select managed identities** panel.
1. On the **Select managed identities** panel, use the **Subscription** and **Managed identity** dropdowns to filter the search results for your identities. Use the **Select** search box to locate the user-assigned managed identity you enabled for the Azure resource hosting your app.

    :::image type="content" source="../media/user-assigned-identity-assign-roles.png" alt-text="A screenshot showing the managed identity assignment process.":::

1. Select the identity and choose **Select** at the bottom of the panel to continue.
1. Select **Review + assign** at the bottom of the page.
1. On the final **Review + assign** tab, select **Review + assign** to complete the workflow.

### [Azure CLI](#tab/azure-cli)

> To assign a role to a user-assigned identity using the Azure CLI, you'll need the principal ID of the identity. Use the `az identity show` command to retrieve the resource ID:

Reviewer: ```suggestion
To assign a role to a user-assigned managed identity using the Azure CLI, you'll need the principal ID of the identity. Use the `az identity show` command to retrieve the resource ID:
```

1. On the **Managed Identities** page, select **+ Create**.

    :::image type="content" source="../media/user-assigned-identity-create.png" alt-text="A screenshot showing the page to manage user-assigned identities.":::

1. On the **Create User Assigned Managed Identity** page, select a subscription, resource group, and region for the user-assigned managed identity, and then provide a name.
1. Select **Review + create** to review and validate your inputs.

    :::image type="content" source="../media/user-assigned-identity-form.png" alt-text="A screenshot showing the form to create a user-assigned managed identity.":::

1. Select **Create** to create the user-assigned managed identity.
1. After the identity is created, select **Go to resource**.
1. On the new identity's **Overview** page, copy the `Client ID` value to use for later when you configure the application code.

### [Azure CLI](#tab/azure-cli)

Use the Azure CLI command [`az identity create`](/cli/azure/identity?view=azure-cli-latest#az-identity-create) to create a managed identity:

```azurecli
az identity create \
    --resource-group <resource-group-name> \

>     --name <identity-name>

Reviewer: ```suggestion
    --name <identity-name> \
    --query 'clientId' \
    --output json
```

```

> The command output prints the following values:

Author: I originally defined these to give context to the user in case they start trying to do more with these IDs or if they see the other values in other tools or identity workflows. But I think it's fine to also just keep it simple and let them find these out in other workflows. This is also the type of thing that could go in the overview doc or somewhere else.

- **ClientID**: Used to configure application code that uses the identity.
- **Location**: The Azure region that contains the identity.
- **Name**: The name of the identity.
- **PrincipalId**: Used for access control and role assignments in Azure.
- **ResourceGroup**: The resource group that contains the identity.
- **TenantId**: The Microsoft Entra tenant that contains the identity.

You can always view the managed identity properties again using the [`az identity show`](/cli/azure/identity?view=azure-cli-latest#az-identity-show) command:

```azurecli
az identity show \
    --resource-group <your-resource-group> \
    --name <your-managed-identity-name> \

>     -o json

Reviewer: ```suggestion
    --output json
```

1. Select **+ Add** to open the **Add user assigned managed identity** panel.
1. On the **Add user assigned managed identity** panel, use the **Subscription** dropdown to filter the search results for your identities. Use the **User assigned managed identities** search box to locate the user-assigned managed identity you enabled for the Azure resource hosting your app.
1. Select the identity and choose **Add** at the bottom of the panel to continue.

    :::image type="content" source="../media/add-user-assigned-identity-to-app.png" alt-text="A screenshot showing how to associate a user-assigned identity with an app.":::

### [Azure CLI](#tab/azure-cli)

The Azure CLI provides different commands to assign a user-assigned identity to different types of hosting services.

To assign a user-assigned identity to a resource such as an Azure App Service web app using the Azure CLI, you'll need the resource ID of the identity. Use the [`az identity show`](/cli/azure/identity?view=azure-cli-latest#az-identity-show) command to retrieve the resource ID:

```azurecli
az identity show \
    --resource-group <your-resource-group> \
    --name <your-managed-identity-name> \

>     -o json --query id

Reviewer: ```suggestion
    --output json \
    --query id
```

1. For the **Assign access to** option, select **Managed identity**.
1. For the **Members** option, choose **+ Select members** to open the **Select managed identities** panel.
1. On the **Select managed identities** panel, use the **Subscription** and **Managed identity** dropdowns to filter the search results for your identities. Use the **Select** search box to locate the user-assigned managed identity you enabled for the Azure resource hosting your app.

    :::image type="content" source="../media/user-assigned-identity-assign-roles.png" alt-text="A screenshot showing the managed identity assignment process.":::

1. Select the identity and choose **Select** at the bottom of the panel to continue.
1. Select **Review + assign** at the bottom of the page.
1. On the final **Review + assign** tab, select **Review + assign** to complete the workflow.

### [Azure CLI](#tab/azure-cli)

To assign a role to a user-assigned identity using the Azure CLI, you'll need the principal ID of the identity. Use the `az identity show` command to retrieve the resource ID:

```azurecli
az identity show \
    --resource-group <your-resource-group> \
    --name <your-managed-identity-name> \

>     -o json --query principalId

Reviewer: ```suggestion
    --output json \
    --query principalId
```
